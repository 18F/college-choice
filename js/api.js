---
# // with frontmatter, we can use {{ variable }} template tags
---
(function(exports) {
  
  var API = {
    url: '{{ site.api.baseurl }}'
  };

  API.get = function(uri, params, done) {
    console.debug('[API] get("%s", %s)', uri, JSON.stringify(params));
    if (arguments.length === 2) {
      done = params;
    } else if (params) {
      if (typeof params === 'object') {
        params = querystring.stringify(params);
      }
      uri = join([uri, params], '?');
    }
    var url = join([API.url, uri], '/');
    console.debug('[API] getting: "%s"', url);
    return d3.json(url, done);
  };

  API.load = function(uri, done) {
    var ext = uri.split('.').pop();
    var load = d3[ext || 'json'];
    return load(uri, done);
  };

  API.endpoint = function(uri) {
    return function endpoint(params, done) {
      return API.get(uri, params, done);
    };
  };

  var schoolEndpoint = 'school/';
  var idField = 'id';

  API.search = API.endpoint(schoolEndpoint);

  API.getSchool = function(id, done) {
    var data = {};
    data[idField] = id;
    return API.get(schoolEndpoint, data, function(error, schools) {
      return error || !schools.length
        ? done(error ? error.responseText : 'No such school found.')
        : done(null, schools[0]);
    });
  };

  exports.API = API;

  function join(list, glue) {
    for (var i = 0; i < list.length; i++) {
      var str = String(list[i]);
      if (str.charAt(0) === glue) {
        list[i] = str.substr(1);
      } else if (str.charAt(str.length - 1) === glue) {
        list[i] = str.substr(0, str.length - 1);
      }
    }
    return list.join(glue);
  }

})(this);
