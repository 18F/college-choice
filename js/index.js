(function(exports) {

  var form = new formdb.Form('#search-form');
  picc.form.minifyQueryString(form);

})(this);
