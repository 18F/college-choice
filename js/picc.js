---
# // with frontmatter, we can use {{ variable }} template tags
---
(function(exports) {

  var picc = exports.picc = {};

  picc.BASE_URL = '{{ site.baseurl }}';

  picc.API = (function() {
    var API = {
      url: '{{ site.api.baseurl }}',
      key: '{{ site.api.key }}'
    };

    // use the staging API if we're on Federalist previewing
    // the staging branch
    if (location.hostname === 'federalist.18f.gov'
        && !!location.pathname.match(/\/staging\//)) {
      API.url = 'https://ccapi-open-staging.18f.gov/';
      API.key = '';
    }

    var schoolEndpoint = 'school/';
    var idField = 'id';

    API.get = function(uri, params, done) {
      // console.debug('[API] get("%s", %s)', uri, JSON.stringify(params));
      if (arguments.length === 2) {
        done = params;
      } else if (params) {
        if (typeof params === 'object') {
          if (API.key) params.api_key = API.key;
          params = querystring.stringify(params);
        } else if (API.key) {
          params += '&api_key=' + API.key;
        }
        uri = join([uri, params], '?');
      }
      var url = join([API.url, uri], '/');
      console.debug('[API] get: "%s"', url);
      return d3.json(url, done);
    };

    API.load = function(uri, done) {
      var ext = uri.split('.').pop();
      var load = d3[ext || 'json'];
      return load(uri, done);
    };

    API.endpoint = function(uri) {
      return function endpoint(params, done) {
        return API.get(uri, params, done);
      };
    };

    API.search = API.endpoint(schoolEndpoint);

    API.getSchool = function(id, done) {
      var data = {};
      data[idField] = id;
      return API.get(schoolEndpoint, data, function(error, res) {
        if (error || !res.total) {
          return done(error.responseText || 'No such school found.');
        } else if (res.total > 1) {
          console.warn('More than one school found for ID: "' + id + '"');
        }
        return done(null, res.results[0]);
      });
    };

    API.getAll = function(urls, done) {
      Object.keys(urls).forEach(function(key) {
        var url = urls[key];
        urls[key] = Array.isArray(url)
          ? function(done) {
            var method = url.shift();
            if (typeof method === 'string') {
              method = API[method];
            }
            url.push(done);
            return method.apply(API, url);
          }
          : function(done) {
            return API.get(url, done);
          };
      });
      // console.log('getAll:', urls);
      return async.parallel(urls, done);
    };

    function join(list, glue) {
      for (var i = 0; i < list.length; i++) {
        var str = String(list[i]);
        if (str.charAt(0) === glue) {
          list[i] = str.substr(1);
        } else if (str.charAt(str.length - 1) === glue) {
          list[i] = str.substr(0, str.length - 1);
        }
      }
      return list.join(glue);
    }

    return API;
  })();

  /*
   * This is a dictionary for the various "special designation"
   * columns. The race/ethnicity ones are all nested under the
   * `minority_serving` property, whereas the `women_only` and
   * `men_only` are top-level properties of each school API response
   * object.
   */
  var SPECIAL_DESIGNATIONS = {
    // TODO: rename 'aanapi' or 'aanapisi'
    // per <http://www2.ed.gov/programs/aanapi/index.html>
    aanipi:               'AANAPI',
    hispanic:             'Hispanic',
    historically_black:   'Historically Black',
    predominantly_black:  'Predominantly Black',
    tribal:               'Tribal',
    women_only:           'Women Only',
    men_only:             'Men Only'
  };

  var NA = '--';

  /**
   * This is our format generator. Its methods are format generators for
   * specific types of values, and they take a key in the data object to
   * format. For instance:
   *
   * @example
   * var formatFoo = format.percent('foo');
   * formatFoo({foo: 0.5}) -> '50%'
   */
  picc.format = (function() {
    var formatter = function(fmt, _empty) {
      if (typeof fmt === 'string') {
        fmt = d3.format(fmt);
      }
      return function(key, empty) {
        empty = empty || _empty;
        if (typeof empty === 'string') {
          empty = d3.functor(empty);
        }
        key = key
          ? picc.access(key)
          : function(v) { return v; };
        return function(d) {
          var value = key.call(this, d);
          return ((value === '' || isNaN(value)) && empty)
            ? empty.call(d)
            : fmt.call(d, +value, key);
        };
      };
    };

    var map = function(keys, fallback) {
      return function(key) {
        return keys[key] || fallback;
      };
    };

    var range = function(ranges) {
      var len = ranges.length;
      var i;
      return function(value) {
        value = +value;
        for (i = 0; i < len; i++) {
          var range = ranges[i];
          if (value >= range[0] && value < range[1]) {
            return range[2];
          }
        }
      };
    };

    return {
      // format.dollars('x')({x: 1000}) === '$1,000'
      dollars: formatter('$,d', NA),
      // format.percent('y')({x: 1000}) === '$1,000'
      percent: formatter('%.0f', NA),
      number: formatter(',d', NA),

      // format.plural('x', 'foo')({x: 1}) === 'foo'
      // format.plural('x', 'foo')({x: 2}) === 'foos'
      plural: function(key, singular, plural) {
        if (!plural) plural = singular + 's';
        return function(d) {
          return d[key] == 1 ? singular : plural;
        };
      },

      // format.map('x', {1: 'a'})({x: 1}) === 'a'
      map: function(key, values, empty) {
        return formatter(map(values))
          .call(this, key, empty);
      },

      // format.control('control')({control: 1}) === 'Public'
      // format.control('control')({control: 2}) === 'Private non-profit'
      control: formatter(map({
        '1': 'Public',
        '2': 'Private non-profit',
        '3': 'Private for-profit'
      }, 'control unknown')),

      // format.preddeg('deg')({deg: 2}) === '2-year'
      // format.preddeg('deg')({deg: 3}) === '4-year'
      preddeg: formatter(map({
        '1': 'Certificate',
        '2': '2-year',
        '3': '4-year',
        '4': 'Graduate'
      }, NA)),

      sizeCategory: formatter(range([
        [0, 2000, 'Small'],
        [2000, 15000, 'Medium'],
        [15000, Infinity, 'Large']
      ]), 'size unknown'),

      // format.locale('locale')({locale: 11}) === 'City: Large'
      locale: formatter(map({
        '11': 'City: Large',
        '12': 'City: Midsize',
        '13': 'City: Small',
        '21': 'Suburb: Large',
        '22': 'Suburb: Midsize',
        '23': 'Suburb: Small',
        '31': 'Town: Fringe',
        '32': 'Town: Distant',
        '33': 'Town: Remote',
        '41': 'Rural: Fringe',
        '42': 'Rural: Distant',
        '43': 'Rural: Remote'
      }, 'locale unknown'))

    };
  })();

  picc.access = function(key) {
    return (typeof key === 'function')
      ? key
      : getter(key);
  };

  function getter(key) {
    if (key.indexOf('.') > -1) {
      var bits = key.split('.');
      var len = bits.length;
      return function(d) {
        for (var i = 0; i < len; i++) {
          d = d[bits[i]];
        }
        return d;
      };
    }
    return function(d) { return d[key]; };
  }

  picc.access.publicPrivate = function(d) {
    switch (+d.ownership) {
      case 1: // public
        return 'public';

      case 2: // private
      case 3:
        return 'private';
    }
    return null;
  };

  picc.access.nationalStat = function(stat, suffix) {
    if (suffix) {
      suffix = picc.access(suffix);
      return function(d) {
        var key = suffix.apply(this, arguments);
        return this.getAttribute([
          'data', stat, key
        ].join('-'));
      };
    } else {
      return function() {
        return this.getAttribute('data-' + stat);
      };
    }
  };

  picc.access.netPrice = function(d) {
    var key = picc.access.publicPrivate(d);
    return key
      ? d.avg_net_price
        ? picc.nullify(d.avg_net_price[key])
        : picc.nullify(d.net_price[key].average)
      : null;
  };

  picc.access.netPriceByIncomeLevel = function(level) {
    return function(d) {
      var key = picc.access.publicPrivate(d);
      return d.net_price
        ? picc.nullify(d.net_price[key].by_income_level[level])
        : null;
    };
  };

  picc.access.yearDesignation = function(d) {
    switch (d.common_degree) {
      case '2': // 2-year (AKA less than 4-year)
        return 'lt_four_year';
      case '3': // 4-year
        return 'four_year';
    }
    // FIXME
    return 'other';
  };

  picc.access.earningsMedian = function(d) {
    return picc.nullify(d.earnings
      ? d.earnings.median
      : d.median_earnings);
  };

  picc.access.earnings25k = function(d) {
    return d.earnings
      ? picc.nullify(d.earnings.percent_gt_25k)
      : null;
  };

  picc.access.completionRate = function(d) {
    var designation = picc.access.yearDesignation(d);
    return designation
      ? picc.nullify(d.completion_rate[designation])
      : null;
  };

  picc.access.partTimeShare = function(d) {
    if (d.part_time_share) {
      var share = picc.nullify(d.part_time_share[1]);
      return share === null ? null : Math.round(+share * 100);
    }
    return null;
  };

  picc.access.retentionRate = function(d) {
    var designation = picc.access.yearDesignation(d);
    // FIXME: use partTimeShare() accessor?
    var partTimeShare = +d.part_time_share[1] || +d.part_time_share[2];
    var retention = d.retention_rate[designation];
    var partTimeRate = retention ? retention.part_time : 0;
    var fullTime = retention ? retention.full_time : 0;
    var size = d.size;
    return (
      (size * partTimeShare * partTimeRate) +
      ((size - size * partTimeShare) * fullTime)
    ) / size;
  };

  picc.access.specialDesignations = function(d) {
    var designations = [];

    if (+d.women_only) {
      designations.push(SPECIAL_DESIGNATIONS.women_only);
    } else if (+d.men_only) {
      designations.push(SPECIAL_DESIGNATIONS.men_only);
    }

    if (d.minority_serving) {
      for (var key in SPECIAL_DESIGNATIONS) {
        if (+d.minority_serving[key]) {
          designations.push(SPECIAL_DESIGNATIONS[key]);
        }
      }
    }

    return designations;
  };

  picc.access.programAreas = function(d, metadata) {
    if (!metadata) metadata = d.metadata;
    if (!metadata || !metadata.dictionary) return [];

    var dictionary = metadata.dictionary;
    return Object.keys(d.program_percentage || {})
      .map(function(key) {
        var value = d.program_percentage[key];
        var dictKey = 'program_percentage.' + key;
        var name = dictionary[dictKey]
          ? dictionary[dictKey].description
          : key;
        return {
          program:  name,
          percent:  value
        };
      });
  };

  picc.nullify = function(value) {
    return value === 'NULL' ? null : value;
  };

  /**
   * namespace for school-related stuff
   */
  picc.school = {};

  /**
   * common directives for school templates
   */
  picc.school.directives = (function() {
    var access = picc.access;
    var format = picc.format;
    var percent = format.percent();

    var href = function(d) {
      var name = d.name.replace(/\W+/g, '-');
      return [
        picc.BASE_URL, '/school/?',
        d.id, '-', name
      ].join('');
    };

    return {
      title: {
        link: {
          text: 'name',
          '@href': href
        }
      },

      size_number:    format.number('size'),
      control:        format.control('ownership'),
      locale_name:    format.locale('locale'),
      years:          format.preddeg('common_degree'),
      size_category:  format.sizeCategory('size'),

      // this is a direct accessor because some designations
      // (e.g. `women_only`) are at the object root, rather than
      // nested in `minority_serving`.
      special_designations: access.specialDesignations,

      SAT_avg: function(d) {
        return picc.nullify(d.SAT_avg) || NA;
      },

      SAT_meter: {
        // TODO
      },

      ACT_meter: {
        // TODO
      },

      average_cost: format.dollars(access.netPrice),
      average_cost_meter: {
        '@max':     access.nationalStat('max', access.publicPrivate),
        '@average': access.nationalStat('median', access.publicPrivate),
        '@value':   access.netPrice,
        label:      format.dollars(access.nationalStat('median', access.publicPrivate)),
        '@title':   debugMeterTitle
      },

      // income level net price stats
      net_price_income1: format.dollars(access.netPriceByIncomeLevel('0-30000')),
      net_price_income2: format.dollars(access.netPriceByIncomeLevel('30001-48000')),
      net_price_income3: format.dollars(access.netPriceByIncomeLevel('48001-75000')),
      net_price_income4: format.dollars(access.netPriceByIncomeLevel('75001-110000')),
      net_price_income5: format.dollars(access.netPriceByIncomeLevel('110001-plus')),

      grad_rate: format.percent(access.completionRate),
      grad_rate_meter: {
        '@average': access.nationalStat('median', access.yearDesignation),
        '@value':   access.completionRate,
        label:      format.percent(function() {
          return this.getAttribute('average');
        }),
        '@title':   debugMeterTitle
      },

      average_salary: format.dollars(access.earningsMedian),
      average_salary_meter: {
        '@value': access.earningsMedian,
        label:    format.dollars(function() {
          return this.getAttribute('average');
        }),
        '@title': debugMeterTitle
      },

      repayment_rate_percent: format.percent('repayment_rate'),
      repayment_rate_meter: {
        '@value': 'repayment_rate',
        '@average': access.nationalStat('median'),
        label:    format.percent(function() {
          return this.getAttribute('average');
        })
      },

      average_total_debt: format.dollars('debt.median_completer_total'),
      average_monthly_loan_payment: format.dollars('debt.median_monthly_payment'),

      federal_aid_percentage: format.percent(function(d) {
        if (!d.loan_rate) return null;
        return Math.max(
          picc.nullify(d.loan_rate.federal),
          picc.nullify(d.loan_rate.pell)
        ) || null; // 0 is n/a
      }),

      earnings_gt_25k: format.percent(access.earnings25k),
      earnings_gt_25k_meter: {
        '@value': access.earnings25k,
        label: format.percent(function() {
          return this.getAttribute('average');
        }),
        '@title': debugMeterTitle
      },

      retention_rate_value: format.percent(picc.access.retentionRate),
      retention_rate_meter: {
        '@value': access.retentionRate,
        label:    format.percent(function() {
          return this.getAttribute('average');
        }),
        '@title': debugMeterTitle
      },

      full_time_percent: format.number(function(d) {
        var pt = access.partTimeShare(d);
        return pt === null ? null : (100 - pt);
      }),

      part_time_percent: format.number(access.partTimeShare),

      gender_values: function(d) {
        if (!d.demographics) return [];
        var female = picc.nullify(d.demographics.female_percent);
        if (female === null) return [];
        female = +female;
        return [
          {label: 'Female', value: female, percent: percent(female)},
          {label: 'Male', value: 1 - female, percent: percent(1 - female)}
        ];
      },

      race_ethnicity_values: function(d) {
        if (!d.demographics || !d.metadata) {
          console.warn('no demographics or metadata:', d);
          return [];
        }

        var dictionary = d.metadata.dictionary;
        var values = d.demographics.race_ethnicity;
        var prefix = 'demographics.race_ethnicity.';
        return Object.keys(values)
          .map(function(key) {
            var value = picc.nullify(values[key]);
            var dict = dictionary[prefix + key];
            return {
              key: key,
              label: dict ? dict.label : key,
              value: value,
              percent: percent(value)
            };
          })
          .filter(function(d) {
            return d.value > 0;
          });
      },

      available_programs: function(d) {
        var areas = access.programAreas(d);
        return areas
          .sort(function(a, b) {
            return d3.ascending(a.program, b.program);
          });
      },

      num_available_programs: format.number(function(d) {
        return access.programAreas(d).length;
      }),

      popular_programs: function(d) {
        var areas = access.programAreas(d);
        if (areas.length) {
          var total = d3.sum(areas, picc.access('percent'));
          areas.forEach(function(d) {
            d.value = +d.percent / total;
            d.percent = percent(d.value);
          });
        }
        return areas
          .sort(function(a, b) {
            return d3.descending(a.value, b.value);
          })
          .slice(0, 5);
      },

      age_entry: function(d) {
        return d.demographics
          ? picc.nullify(d.demographics.age_entry)
          : null;
      },

      more_link: {
        '@href': href
      }
    };

    function debugMeterTitle(d) {
      return [
        'value: ', this.getAttribute('value'), '\n',
        'median: ', this.getAttribute('average')
      ].join('');
    }

  })();


  // form utilities
  picc.form = {};

  /**
   * Adds a "submit" listener to the provided formdb.Form
   * instance (or CSS selector) that intercepts its data,
   * formats it as a querystring, then does a client-side
   * redirect with window.location, effectively removing
   * the query string parameters for empty inputs.
   */
  picc.form.minifyQueryString = function(form) {

    // allow form to be a CSS selector
    if (typeof form !== 'object') {
      form = new formdb.Form(form);
    }

    form.on('submit', function(data, e) {
      var url = [
        form.element.action,
        querystring.stringify(data)
      ].join('?');

      window.location = url;
      e.preventDefault();
      return false;
    });

    return form;
  };


  // UI tools
  picc.ui = {};

  picc.ui.expandAccordions = function(selector, expanded) {
    if (arguments.length === 1) {
      expanded = selector;
      selector = null;
    }
    if (!selector) {
      selector = 'aria-accordion';
    }
    expanded = d3.functor(expanded);
    return d3.selectAll(selector)
      .filter(function() {
        return !!expanded.apply(this, arguments);
      })
      .property('expanded', true);
  };

  // debounce function
  picc.debounce = function(fn, delay) {
    var timeout;
    return function() {
      var context = this;
      var args = arguments;
      return timeout = setTimeout(function() {
        fn.apply(context, args);
      }, delay);
    };
  };

})(this);
