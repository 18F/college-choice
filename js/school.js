---
# // hey there
---
(function(exports) {

  var id = getSchoolId();

  if (!id) {
    return showError('No school ID provided');
  }

  var root = document.querySelector('#school');

  var format = picc.format;

  var directives = {
    size_number: format.number('size'),
    control: format.control('ownership'),
    locale_name: format.locale('locale'),
    years: format.preddeg('common_degree'),
    size_category: format.sizeCategory('size'),

    // this is a direct accessor because some designations
    // (e.g. `women_only`) are at the object root, rather than
    // nested in `minority_serving`.
    special_designation: format.specialDesignation,

    average_cost: format.dollars('avg_net_price'),
    average_cost_meter: {
      '@max':     d3.functor(100000),
      '@average': d3.functor(20000),
      '@value':   function(d) {
        return Math.floor(Math.random() * 100000);
      }
    },

    // for debugging
    map: {
      '@data-latitude': function(d) {
        return d.location.lat;
      },
      '@data-longitude': function(d) {
        return d.location.lon;
      }
    }
  };

  picc.API.getSchool(id, function(error, school) {
    if (error) {
      return showError(error);
    }

    console.log('got school:', school);
    root.classList.remove('hidden');
    tagalong(root, school, directives);

    var center = L.latLng(
      +school.location.lat,
      +school.location.lon
    );

    var map = L.map(root.querySelector('.school-map'), {
        zoomControl:        false,
        panControl:         false,
        attributionControl: false,
        dragging:           false,
        scrollWheelZoom:    false,
        touchZoom:          false,
        doubleClickZoom:    false,
        boxZoom:            false
      })
      .setView(center, 13);

    L.tileLayer('http://tile.stamen.com/terrain/{z}/{x}/{y}.png')
      .addTo(map);

    var marker = L.circle(center, 200, {
        color: 'black',
        opacity: 1,
        strokeWidth: 1,
        fillColor: 'white',
        fillOpacity: 1
      })
      .addTo(map);

    marker.bindPopup(school.name);
  });

  function getSchoolId() {
    if (!location.search) return null;
    var match = location.search.match(/^\?(\d+)(\b|-)/);
    return match ? match[1] : null;
  }

  function redirect(uri) {
    window.location = uri || '../search/';
  }

  function showError(message) {
    var container = document.querySelector('#error');
    container.classList.remove('hidden');
    var target = container.querySelector('.error-message') || container;
    target.textContent = message;
    return target;
  }

})(this);
