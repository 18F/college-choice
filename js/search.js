(function(exports) {

  var resultsRoot = document.querySelector('.search-results');
  var query = location.search;
  if (!query) {
    resultsRoot.classList.add('hidden');
    return;
  }

  // TODO: include formdb, querystring, and tagalong
  var form = new formdb.Form('#search-form');
  var values = querystring.parse(query.substr(1));
  removeEmptyValues(values);
  console.debug('[search] form values:', values);
  form.setData(values);

  expandAccordions(values);

  var format = picc.format;

  resultsRoot.classList.add('js-loading');
  picc.API.search(values, function(error, data) {
    resultsRoot.classList.remove('hidden');
    resultsRoot.classList.remove('js-loading');

    if (error) {
      return showError(error);
    }

    console.log('loaded schools:', data);
    resultsRoot.classList.add('js-loaded');

    console.time('[render]');

    console.time('[render] template');
    // render the basic DOM template for each school
    tagalong(resultsRoot, data, {
      results_word: format.plural('total', 'Result')
    });

    d3.select(resultsRoot)
      .datum(data)
      .select('[data-bind="total"]')
        .text(format.number('total', '0'));

    var resultsList = resultsRoot.querySelector('.schools-list');
    tagalong(resultsList, data.results, {
      size_pretty: format.number('size'),
      link: {
        '@href': function(d) {
          var name = d.name.replace(/\W+/g, '-');
          return ['../school/?', d.id, '-', name].join('');
        }
      }
    });
    console.timeEnd('[render] template');

    console.time('[render] charts');
    // bind all of the data to elements in d3, then
    // call renderCharts() on the selection
    d3.select(resultsList)
      .selectAll('.school')
      .data(data.results)
      .call(renderCharts);
    console.timeEnd('[render] charts');

    console.timeEnd('[render]');
  });

  function renderCharts(selection) {
    selection.select('[data-bind="size"]')
      .text(format.number('size'));
  }

  function showError(error) {
    console.error('error:', error);
    resultsRoot.classList.add('js-error');
    var out = resultsRoot.querySelector('.error');
    out.classList.remove('hidden');
    out.textContent = String(error);
  }

  function removeEmptyValues(obj) {
    var empty = function(v) {
      return v === null || v === '';
    };
    for (var key in obj) {
      if (empty(obj[key])) {
        delete obj[key];
      }
    }
    return obj;
  }

  function expandAccordions(values) {
    for (var key in values) {
      if (values[key]) {
        var el = form.getInputsByName(key)[0];
        if (!el) {
          console.warn('no input element for parameter: "%s"', key);
          continue;
        }
        do {
          el = el.parentNode;
          if (el.classList.contains('picc-accordion')) {
            d3.select(el)
              .attr('data-expanded', 'true');
            break;
          }
        } while (el);
      }
    }
  }

})(this);
