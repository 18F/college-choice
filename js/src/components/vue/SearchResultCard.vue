<style lang="scss">

  .nameLink{
    text-decoration: none;
    line-height: 125%;
  }
  .search-result-card{
    border: 4px solid #FFFFFF !important;
  }
  .result-card-selected{
    border: 4px solid #eeba28 !important;    
  }
</style>

<template>
  <v-card tile class="search-result-card mx-auto pa-0" 
    outlined :class="{'result-card-selected': isSelected}"> <!-- Better Selected style -->
    <v-card-text>
        <p class='mt-1 my-2'>
          <v-chip
          v-if="_.get(school, fields['UNDER_INVESTIGATION'])==1"
          color="error"
          label
          >
          <strong>Under ED Monitoring</strong>
          <tooltip definition="hcm2" color="#FFFFFF" class='ml-2' />
        </v-chip>
        </p>

        <v-btn color="primary" small fab icon ripple class='float-right' :class="{amber: isSelected}" @click="$emit('toggle-compare-school',school)">
          <v-icon>fa fa-star</v-icon>
        </v-btn>
        <p class='overline font-weight-bold mb-1'>{{school['school.city']}}, {{school['school.state']}}</p>
        <h2 class="title mt-0 font-strong"><a class='nameLink' :href="link">{{school['school.name'] ? school['school.name'] : 'School Name'}}</a></h2>
        <p class='subtitle-1 font-italic'>{{school['latest.student.size'] | separator }} undergrads</p>
        <v-divider />
        <v-row> 
        <v-col cols='5'  class="pr-0 text-center">
          <h3>{{displayGradRate  | numeral('0.%') }}</h3>
        </v-col>
        <v-col cols='7'>
          <span>who go graduate <tooltip definition="graduation-rate" /></span>
        </v-col>
      </v-row>
      <v-row class='result-card-info-container'>
        <v-col cols='5' class="pr-0 text-center">
          <h3>{{displayEarn}}</h3>
        </v-col>
        <v-col cols='7'>
          <span>typical earnings for recent graduates <tooltip definition="avg-salary" /></span>          
        </v-col>
      </v-row>
      <v-row class='result-card-info-container' v-if="!isProgramReporter">
        <v-col cols='5' class="pr-0 text-center">
          <h3>{{displayAvgCost}}</h3>
        </v-col>
        <v-col cols='7'>
          <span>average annual cost after aid <tooltip definition="avg-cost-year" /></span>
        </v-col>
      </v-row>
      <v-row class='result-card-info-container' v-else>
        <v-col cols="12" class='text-center'>
          <em>Average annual cost is not available for program reporters.</em>
        </v-col>
      </v-row>
      <v-row>
        <v-col cols="12">
          <small-school-icons :school="school" :fields="fields" size="small" />
        </v-col>
      </v-row>
    </v-card-text>
  </v-card>
</template>

<script>
import numeral from 'numeral';
import Tooltip from 'components/vue/Tooltip.vue';
import SmallSchoolIcons from 'components/vue/SmallSchoolIcons.vue';
export default {
  components: {
    tooltip: Tooltip,
    'small-school-icons': SmallSchoolIcons
  },
  props:{
      "school": Object,
      "isSelected": Boolean
  },
  data(){
    return {
      fields: picc.fields
    }
  },
  computed:{
    link(){
      let id = _.get(this.school, this.fields['ID']);
      let name = _.get(this.school, this.fields['NAME'],'(unknown)');
      return '/school/?'+id+'-'+name.replace(/\W+/g, '-'); 
    },
    displayGradRate(){
      let OM = _.get(this.school, this.fields.COMPLETION_OM);
      let G200_4 = _.get(this.school, this.fields.COMPLETION_200_4);
      let G200_LT4 = _.get(this.school, this.fields.COMPLETION_200_LT4);
      if(OM){
        return OM;
      }
      else{
        if(G200_LT4 || G200_4) return (this.years==3) ? G200_4 : G200_LT4;
        else return "N/A";
      }
    },
    isProgramReporter() {
      return _.get(this.school, this.fields.PROGRAM_REPORTER_OFFERED) > 0;
    },
    displayEarn(){
      // if (!this.school['latest.earnings.10_yrs_after_entry.median'] || this.school['latest.earnings.10_yrs_after_entry.median'] < 0){
      //   return "N/A"
      // }else{
      //   return numeral(this.school['latest.earnings.10_yrs_after_entry.median']).format('$0a');
      // }
      return "$12K - $42K";
    },
    displayAvgCost(){
      if (!this.school['latest.cost.avg_net_price.overall'] || this.school['latest.cost.avg_net_price.overall'] < 0){
        return "N/A"
      }else{
        return numeral(this.school['latest.cost.avg_net_price.overall']).format('$0a');
      }
    }
  }
}
</script>