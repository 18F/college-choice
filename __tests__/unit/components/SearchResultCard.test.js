import {shallowMountWithConfig} from '../../utility/ShallowMount';

import SearchResultCard from '../../../js/src/components/vue/SearchResultCard.vue';
import schoolData from '../../mock/school.json';
import {fields as fieldData} from '../../../js/src/vue/constants.js';

const mountOptions = {
  propsData:{
    school: schoolData,
    fields: fieldData
  }
}

describe('Component: SearchResultCard', () => {

  beforeEach(() => {
  })

  test('is a Vue instance', () => {
    const wrapper = shallowMountWithConfig(SearchResultCard,mountOptions);

    console.log(wrapper.html());
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
})