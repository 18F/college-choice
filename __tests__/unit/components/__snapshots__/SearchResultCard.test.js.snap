// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SearchResultCard Renders correctly 1`] = `
<div
  class="search-result-card mx-auto pa-0 elevation-4 v-card v-card--outlined v-sheet theme--light"
  fields="[object Object]"
>
  <div
    class="v-card__text pa-3"
  >
    <!---->
     
    <button
      class="float-right search-result-card-compare v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default grey--text"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon notranslate fa fa fa fa-plus-circle theme--light"
        />
         
        <span
          class="sr-only"
        >
          Compare
        </span>
      </span>
    </button>
     
    <p
      class="overline font-weight-bold mb-1"
    >
      New Haven, CT
    </p>
     
    <h2
      class="title mt-0 font-weight-bold"
    >
      <a
        class="nameLink"
        href="/school/?130794-Yale-University"
      >
        Yale University
      </a>
    </h2>
     
    <p
      class="body-2 mt-1"
    >
      5,742 undergrads
    </p>
     
    <hr
      aria-orientation="horizontal"
      class="v-divider theme--light"
      role="separator"
    />
     
    <div
      class="row"
    >
      <div
        class="col col-12"
      >
        <div>
          <ul
            class="school-key_figures key-figures ma-0 pa-0 d-flex justify-space-around"
          >
            <span
              class="v-tooltip v-tooltip--bottom"
            >
              <div
                class="v-tooltip__content"
                style="left: 0px; max-width: auto; opacity: 0; top: 12px; z-index: 0; display: none;"
              />
              <li
                class="icon-four"
              >
                <span>
                  Year
                </span>
              </li>
            </span>
             
            <li
              class="icon-private"
            >
              <span>
                Private
              </span>
            </li>
             
            <li
              class="icon-city"
            >
              <span>
                City
              </span>
            </li>
             
            <li
              class="icon-medium"
            >
              <span>
                Medium
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
     
    <hr
      aria-orientation="horizontal"
      class="v-divider theme--light"
      role="separator"
    />
     
    <div
      class="row v-flex align-center"
    >
      <div
        class="py-2 col col-7"
      >
        <span>
          Graduation Rate 
          <span>
            <div
              class="v-dialog__container"
              role="dialog"
            >
              <i
                allow-overflow=""
                class="v-icon notranslate v-icon--link  fas fa-info-circle theme--dark blue--text text--darken-3"
                role="button"
                style="font-size: 12px;"
              />
              <div
                class="v-dialog__content"
                role="document"
                style="z-index: 0;"
              >
                <div
                  class="v-dialog"
                  style="max-width: 500px; display: none;"
                />
              </div>
            </div>
          </span>
        </span>
      </div>
       
      <div
        class="pr-0 text--black py-2 col col-5"
      >
        <h3
          class="navy-text"
        >
          97%
        </h3>
      </div>
    </div>
     
    <div
      class="row result-card-info-container v-flex align-center"
    >
      <div
        class="py-2 col col-7"
      >
        <span>
          Salary After Completing 
          <span>
            <div
              class="v-dialog__container"
              role="dialog"
            >
              <i
                allow-overflow=""
                class="v-icon notranslate v-icon--link  fas fa-info-circle theme--dark blue--text text--darken-3"
                role="button"
                style="font-size: 12px;"
              />
              <div
                class="v-dialog__content"
                role="document"
                style="z-index: 0;"
              >
                <div
                  class="v-dialog"
                  style="max-width: 500px; display: none;"
                />
              </div>
            </div>
          </span>
        </span>
      </div>
       
      <div
        class="pr-0 text--black py-2 col col-5"
      >
        <h3
          class="navy-text"
        >
          $39k-120k
        </h3>
      </div>
    </div>
     
    <div
      class="row result-card-info-container v-flex align-center"
    >
      <div
        class="py-2 col col-7"
      >
        <span>
          Average Annual Cost 
          <span>
            <div
              class="v-dialog__container"
              role="dialog"
            >
              <i
                allow-overflow=""
                class="v-icon notranslate v-icon--link  fas fa-info-circle theme--dark blue--text text--darken-3"
                role="button"
                style="font-size: 12px;"
              />
              <div
                class="v-dialog__content"
                role="document"
                style="z-index: 0;"
              >
                <div
                  class="v-dialog"
                  style="max-width: 500px; display: none;"
                />
              </div>
            </div>
          </span>
        </span>
      </div>
       
      <div
        class="pr-2 text--black py-0 col col-5"
      >
        <h3
          class="navy-text"
        >
          $19k
        </h3>
      </div>
    </div>
  </div>
</div>
`;
